ast/src/analyzed/mod.rs

```diff
@@ -715,96 +710,231 @@ impl<Expr> Default for SelectedExpressions<Expr> {
     }
 }
 
-impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
+impl<T> Children<AlgebraicExpression<T>> for SelectedExpressions<T> {
     /// Returns an iterator over all (top-level) expressions in this SelectedExpressions.
-    fn children(&self) -> Box<dyn Iterator<Item = &Expr> + '_> {
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
         Box::new(self.selector.iter().chain(self.expressions.iter()))
     }
     /// Returns an iterator over all (top-level) expressions in this SelectedExpressions.
-    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut Expr> + '_> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
         Box::new(self.selector.iter_mut().chain(self.expressions.iter_mut()))
     }
 }
 
 #[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize, JsonSchema)]
-pub struct Identity<SelectedExpressions> {
-    /// The ID is globally unique among identities.
+pub struct PolynomialIdentity<T> {
+    // The ID is globally unique among identitites.
     pub id: u64,
-    pub kind: IdentityKind,
     pub source: SourceRef,
-    /// For a simple polynomial identity, the selector contains
-    /// the actual expression (see expression_for_poly_id).
-    pub left: SelectedExpressions,
-    pub right: SelectedExpressions,
+    pub expression: AlgebraicExpression<T>,
 }
 
-// TODO This is the only version of Identity left.
-impl<T> Identity<SelectedExpressions<AlgebraicExpression<T>>> {
-    /// Constructs an Identity from a polynomial identity (expression assumed to be identical zero).
-    pub fn from_polynomial_identity(
+impl<T> Children<AlgebraicExpression<T>> for PolynomialIdentity<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        Box::new(iter::once(&mut self.expression))
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(iter::once(&self.expression))
+    }
+}
+
+impl<T> PolynomialIdentity<T> {
+    pub fn new(id: u64, source: SourceRef, expression: AlgebraicExpression<T>) -> Self {
```

Comment:
Do we acutally need this when the fields are all `pub`?

Reply:
No, personal preference but happy to remove these!

Reply:
removed